File: arim/Week9/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        int answer = 0;
        Arrays.sort(rocks);
        int left=1, right=distance;
        
        while(left<=right){
            int mid=(left+right)/2;
            if(removeCnt(rocks, mid, distance)<=n){
                answer=mid;
                left=mid+1;
            }
            else right=mid-1;
        }
        return answer;
    }
    
    public int removeCnt(int[] rocks, int mid, int distance){
        int before=0, end=distance, cnt=0;
        
        for(int i=0; i<rocks.length; i++){
            if(rocks[i]-before<mid){
                cnt++;
                continue;
            }
            before=rocks[i];
        }
        if(end-before<mid) cnt++;
        
        return cnt;
    }
}

File: 05ovo2e/Week9/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(int distance, vector<int> rocks, int n) {
   
    sort(rocks.begin(), rocks.end());
    // left는 최소 거리, right는 최대 거리(출발지와 도착지의 거리)
    // 중간값 `mid`를 최소 거리로 하여, 
    // 주어진 조건에 맞추어 바위를 제거하는 조건을 확인하면서 left와 right의 값을 조정
    rocks.push_back(distance);
    int left = 1;
    int right = distance;
    long long answer = right;
    
     while (left <= right) {
        int mid = (left + right) / 2; 
        int prev = 0;
        int remove_rock = 0; 
         
         for (int rock : rocks) {
            if (rock - prev < mid) {
                remove_rock++;
            }
            else {
                prev = rock;
            }
        }

        if (remove_rock > n) {
            right = mid - 1;
        } else {
            answer = mid;
            left = mid + 1;
        }
    }
    
    return answer;
}


    